src/main/java/bio/terra/cli/service/utils/HttpUtils.java

```diff
@@ -108,4 +118,78 @@ public static HttpResponse sendHttpRequest(
     // return a POJO that includes both the response body and status code
     return new HttpResponse(responseBody.toString(), statusCode);
   }
+
+  /**
+   * Helper method to do retries. Uses {@link #DEFAULT_MAXIMUM_RETRIES} for maximum number of
+   * retries and {@link #DEFAULT_MILLISECONDS_SLEEP_FOR_RETRY} for the time to sleep between
+   * retries.
+   *
+   * @param makeRequest function to perform the request
+   * @param isRetryable function to test whether the exception is retryable or not
+   * @param <T> type of the Http response (i.e. return type of the makeRequest function)
+   * @return the Http response
+   * @throws Exception if makeRequest throws an exception that is not retryable, or if the maximum
+   *     number of retries was exhausted
+   */
+  public static <T, E extends Exception> T callWithRetries(
+      HttpRequestOperator<T, E> makeRequest, Predicate<Exception> isRetryable)
+      throws E, InterruptedException {
+    return callWithRetries(
+        DEFAULT_MAXIMUM_RETRIES, DEFAULT_MILLISECONDS_SLEEP_FOR_RETRY, makeRequest, isRetryable);
+  }
+  /**
+   * Helper method to do retries.
+   *
+   * @param maxRetries maximum number of times to retry
+   * @param sleepDurationMS number of milliseconds to sleep between tries
+   * @param makeRequest function to perform the request
+   * @param isRetryable function to test whether the exception is retryable or not
+   * @param <T> type of the Http response (i.e. return type of the makeRequest function)
+   * @return the Http response
+   * @throws Exception if makeRequest throws an exception that is not retryable, or if the maximum
+   *     number of retries was exhausted
+   */
+  public static <T, E extends Exception> T callWithRetries(
+      int maxRetries,
+      int sleepDurationMS,
+      HttpRequestOperator<T, E> makeRequest,
+      Predicate<Exception> isRetryable)
+      throws E, InterruptedException {
+    int numTries = 0;
+    Exception lastRetryableException = null;
+    do {
+      numTries++;
+      try {
+        logger.info("Http request attempt #{}", numTries);
+        return makeRequest.makeRequest();
+      } catch (Exception ex) {
+        // if the exception is not retryable, then quit polling
+        if (!isRetryable.test(ex)) {
+          throw ex;
+        } else {
+          // keep track of the last retryable exception so we can re-throw it in case of a timeout
+          lastRetryableException = ex;
+        }
+        logger.info("Caught retryable exception: {}", ex);
+      }
+
+      // sleep before retrying
+      Thread.sleep(sleepDurationMS);
+    } while (numTries <= maxRetries);
+
+    // request with retries timed out
+    throw new RuntimeException(
+        "Http request with retries timed out after " + numTries + " tries.",
+        lastRetryableException);
```

Comment:
I think that instead of wrapping this in a RuntimeException we should throw the last exception.

Reply:
I disagree on this one. I think it's helpful to see that it was a timeout that caused the error, and still include the full exception stack trace.

Reply:
I'll push back a little more. It's always a timeout when we run out of retries. What we want to do after the timeout may be dependent on the exception. When we wrap the error in the runtime exception, the calling code can no longer do anything about the ApiException.